register-machines/src/main.rs

```diff
@@ -13,66 +15,165 @@ enum Instruction {
 
 use Instruction::*;
 fn eval_program(program: &[Instruction], state: &State) -> State {
-    unimplemented!();
+    let (mut l, mut rs) = state.clone();
+    while l < program.len() {
+        match program[l] {
+            Add(i, j) => {
+                *rs.entry(i).or_insert(Godel::new()) += 1;
+                l = j; 
+            },
+            Sub(i, j, k) => {
+                let v = rs.entry(i).or_insert(Godel::new()); 
+                if *v != 0 {
+                    *v -= 1;
+                    l = j;
+                } else {
+                    l = k; 
+                }
+            },
+            Halt => { break; }
+        }
+    }
+    (l, rs)
+}
+fn encode_instruction(ins: &Instruction) -> Godel {
+    match ins {
+        Add(i, j) => 
+            encode_pair1(&Godel::from(2 * i), &Godel::from(*j)),
+        Sub(i, j, k) => 
+            encode_pair1(&Godel::from(2 * i + 1),
+            &encode_pair2(&Godel::from(*j), &Godel::from(*k))
+        ),
+        Halt => Godel::ZERO
+    }
 }
 // <<x,y>> = (2^x)*(2y+1)
-fn encode_pair1(x: u128, y: u128) -> u128 {
-    unimplemented!();
+fn encode_pair1(x: &Godel, y: &Godel) -> Godel {
+    (Godel::from(2) * y + 1) << x.to_u32_wrapping()
 }
```

Comment:
A bit picky, but this does limit the size of x (which easily breaks u32 when programs get large). It doesn't seem like rug supports shifts or pow operations between two Integers, which is a shame really.